### PR TITLE
Action for building docker images automatically

### PR DESCRIPTION
**Pull request by j143**
**Tasks**

- [ ] 1. Run nightly build as cron job
- [x] 2. run test build for each commit
- [ ] 3. version build for release version

**How to Review?**

Published artifact at https://hub.docker.com/r/return01/test-ghactions-systemds

Usage test:

```console
janardhan@cloudshell:~ (ai-experiments-001)$ docker run -it --rm return01/test-ghactions-systemds
Hello from SystemDS
SystemDS Statistics:
Total execution time:           0.020 sec.
```

---
Secrets: `DOCKERHUB_USER` and `DOCKERHUB_TOKEN`
These two variables are already set by INFRA. No need to set from our end.

https://issues.apache.org/jira/browse/INFRA-22430

### CODE REVIEW

**Comment by Baunsgaard** [comment]
I think we can split the images a bit and reduce builds. 

1. Build max once a day. 
2. Only build the test image if we change R dependencies or the pom.xml
3. Only build the other images if there is changes in src. 

Thanks for the help!

File: .github/workflows/docker-pub.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  build:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+  
+    - name: Set up QEMU
```
Comment (Baunsgaard) [inline .github/workflows/docker-pub.yml:20]: QEMU is for ?

File: docker/sysds.Dockerfile [hunk 2]
```diff
@@ -60,14 +60,16 @@ RUN apt-get update -qq \
 	rm -r target/hadoop-test && \
 	rm -r target/maven-archiver && \
 	rm -r target/systemds-** && \
-	rm -r docker && \
+	# rm -r docker && \
```
Comment (Baunsgaard) [inline docker/sysds.Dockerfile:63]: you cannot just comment out a line in this multi-line instruction, then everything after it does not execute.

File: docker/sysds.Dockerfile [hunk 3]
```diff
@@ -60,14 +60,16 @@ RUN apt-get update -qq \
...
 	rm -r src && \
 	rm -r /usr/lib/mvn && \
 	rm -r CONTRIBUTING.md && \
 	rm -r pom.xml && \ 
 	rm -r ~/.m2
 
-COPY docker/mountFolder/main.dml /input/main.dml
+RUN mkdir /input && echo 'print("Hello from SystemDS")' > /input/main.dml
+
+# COPY docker/mountFolder/main.dml /input/main.dml
```
Comment (Baunsgaard) [inline docker/sysds.Dockerfile:73]: i'm all for running it but don't remove the copy, so that the only change you make it  to add the RUN line.

Comment (j143) [reply docker/sysds.Dockerfile:73]: I am having error with running `COPY`

File: .github/workflows/docker-pub.yml (hunk 1 above)
Comment (j143) [reply .github/workflows/docker-pub.yml:20]: It is an add-on, incase we want to build multi arch images. Optional though!

File: docker/sysds.Dockerfile (hunk 2 above)
Comment (j143) [reply docker/sysds.Dockerfile:63]:
```suggestion
```

File: docker/sysds.Dockerfile (hunk 3 above)
Comment (j143) [reply docker/sysds.Dockerfile:73]: This one

image

File: .github/workflows/docker-pub.yml (hunk 1 above)
Comment (Baunsgaard) [reply .github/workflows/docker-pub.yml:20]: hmm, then for now lets not include it if it is not needed

File: docker/sysds.Dockerfile (hunk 2 above)
Comment (Baunsgaard) [reply docker/sysds.Dockerfile:63]: Not sure what you mean?
just for information the reason why i have this line is to remove the docker folder from the internal copied systemds repository, it does not make a difference for the building of the docker image except it makes the end result slightly smaller.

so therefore, don't out comment it :)

File: docker/sysds.Dockerfile (hunk 3 above)
Comment (Baunsgaard) [reply docker/sysds.Dockerfile:73]: this is happening because of your out commented line mentioned above.

Comment (Baunsgaard) [reply docker/sysds.Dockerfile:73]: this one:

```
# rm -r docker && \
```

**Comment by Baunsgaard** [comment]
I think it is fine if we just add the on commit part of the docker for now, 
If it is still working, we should start merging it in.

**Comment by Baunsgaard** [comment]
@j143 
Is the only thing missing not just a Infra ticket to add keys to push the images?

**Comment by j143** [comment]
> Is the only thing missing not just a Infra ticket to add keys to push the images?

There are already secrets available.

the problem seems to be `COPY docker/mountFolder/main.dml /input/main.dml` statement in the Dockerfile.

**Pull request merged**